src/tilerange.jl

```diff
@@ -0,0 +1,222 @@
+
+Construct a sliding tile along range `r` with fixed sliding stride `Δ` and tile length `n`.
+
+# Arguments
+
+- `r`: a range, `CartesianIndices` or `Vector`
+- `n::Integer`: The length of each tile. If keyword `keep_last=true`, the last tile length might be
+  less than `n`.
+- `Δ::Union{Integer, CartesianIndex{1}}=n`: The sliding stride `Δ` is defined as `first(r[n]) -
+  first(r[n-1])`. Using a stride `Δ<n` means there are overlaps between each adjacent tile. `Δ` can
+  also be `CartesianIndex{1}`.
+- `keep_last=true` (keyword): this keyword affects the cases when the last tile has few elements
+  than `n`, in which case, `true`/`false` tells `FixedTileRange` to keep/discard the last tile.
+
+# Examples
+
+```jldoctest; setup=:(using TiledIteration)
+julia> FixedTileRange(2:10, 3)
+3-element FixedTileRange{UnitRange{Int64},Int64,UnitRange{Int64}}:
+ 2:4
+ 5:7
+ 8:10
+
+julia> FixedTileRange(1:10, 4)
+3-element FixedTileRange{UnitRange{Int64},Int64,UnitRange{Int64}}:
+ 1:4
+ 5:8
+ 9:10
+
+julia> FixedTileRange(1:10, 4, 2)
+4-element FixedTileRange{UnitRange{Int64},Int64,UnitRange{Int64}}:
+ 1:4
+ 3:6
+ 5:8
+ 7:10
+
+julia> FixedTileRange(1:10, 4; keep_last=false)
+2-element FixedTileRange{UnitRange{Int64},Int64,UnitRange{Int64}}:
+ 1:4
+ 5:8
+
+julia> FixedTileRange(1:2:10, 4)
+2-element FixedTileRange{StepRange{Int64,Int64},Int64,StepRange{Int64,Int64}}:
+ 1:2:7
+ 5:2:9
+```
+
+Besides an `AbstractUnitRange`, the input range `r` can also be a `CartesianIndices{1}` or more
+generally, an `AbstractVector{<:Integer}`:
+
+```jldoctest; setup=:(using TiledIteration)
+julia> FixedTileRange(CartesianIndices((1:10, )), 4)
+3-element FixedTileRange{CartesianIndices{1,Tuple{UnitRange{Int64}}},Int64,CartesianIndices{1,Tuple{UnitRange{Int64}}}}:
+ [CartesianIndex(1,), CartesianIndex(2,), CartesianIndex(3,), CartesianIndex(4,)]
+ [CartesianIndex(5,), CartesianIndex(6,), CartesianIndex(7,), CartesianIndex(8,)]
+ [CartesianIndex(9,), CartesianIndex(10,)]
+```
+
+!!! warning
+    It usually has bad indexing performance if `r` is not lazily evaluated. For example,
+    `FixedTileRange(collect(1:10), 4)` creates a new `Vector` of length `4` everytime when
+    `getindex` is called.
+"""
+struct FixedTileRange{R, T, RP} <: AbstractTileRange{R}
+    parent::RP
+    n::T
+    Δ::T
+    keep_last::Bool
+
+    # keep `length` information to avoid unnecessary calculation and thus is more performant
+    length::T
+
+    function FixedTileRange(parent::R, n::T, Δ; keep_last::Bool=true) where {R<:Range1, T}
+        _length = _fixedtile_length(parent, n, Δ, keep_last)
+        new{_eltype(R), T, R}(parent, n, Δ, keep_last, _length)
+    end
+end
+FixedTileRange(r::Range1, n::Integer; kwargs...) = FixedTileRange(r, n, n; kwargs...)
+
+_eltype(::Type{R}) where R<:OrdinalRange = StepRange{eltype(R), eltype(R)}
+_eltype(::Type{R}) where R<:AbstractUnitRange = UnitRange{eltype(R)}
+_eltype(::Type{R}) where R<:AbstractVector = R # this includes CartesianIndices{1}
```

Comment:
Yes, `Vector{Int}` counts even though it is not memory-efficient. I choose to support that because `Vector{Int}` is a valid indices type.

Supporting this enables us to pass lazy evaluated vectors, even though I'm not sure if there's practical usage right now. Currently, I'm mostly concerned about CartesianIndices and step range.